itest/lnd_coop_close_external_delivery_test.go

```diff
@@ -0,0 +1,92 @@
+package itest
+
+import (
+	"testing"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lntest"
+	"github.com/stretchr/testify/require"
+)
+
+func testCoopCloseWithExternalDelivery(ht *lntest.HarnessTest) {
+	ht.Run("set delivery address at open", func(t *testing.T) {
+		tt := ht.Subtest(t)
+		testCoopCloseWithExternalDeliveryImpl(tt, true)
+	})
+	ht.Run("set delivery address at close", func(t *testing.T) {
+		tt := ht.Subtest(t)
+		testCoopCloseWithExternalDeliveryImpl(tt, false)
+	})
```

Comment:
Interesting choice, why this and not TDD?

Reply:
What do you mean by TDD? I'm only familiar with the acronym as Test Driven Development, and that is exactly what I've done here.

Reply:
Oh, I was referring to something like this:
![image](https://github.com/lightningnetwork/lnd/assets/59079323/37bb474e-12cf-43ba-b11b-aa1e61158ba2)

source: https://golang.cafe/blog/golang-table-test-example

In the code being reviewed above, a number of things were rewritten twice.

Reply:
I stole the test vector approach from @ellemouton 's refactoring of one of my earlier itests here:

https://github.com/lightningnetwork/lnd/blob/master/itest/lnd_coop_close_with_htlcs_test.go#L24-L34

I also recall @yyforyongyu saying that on a go-forward basis we should discourage table tests like in the image you linked. I tend to agree with this assessment but also would defer this type of decision to the elder wizards. I don't have a principled opinion on how to code test vectors.

Reply:
Just my two cents,

For this case, I would do `testCoopCloseWithExternalDeliveryImpl` and `testCoopCloseWithExternalDeliveryImplWithUpfronShutdown` instead.

I like using table-driven tests when the outputs of a function are only determined by one or two inputs in a test setup, like this [one](https://github.com/lightningnetwork/lnd/blob/master/buffer/buffer_test.go#L35) or this [one](https://github.com/lightningnetwork/lnd/blob/master/lnwire/writer_test.go#L464). They share common heuristics like,
- the method being tested is relatively simple - simple as in the number of inputs it takes and the number of outputs it returns, lines of code, logic, etc.
- the test itself is straightforward, which is usually not the case in itest as we often need a somewhat complicated setup to reach one or multiple states before the actual test.
- the assertions used in the test are **NOT** conditional based on the values of the table. An example I want to discourage is the [multi-hop itest](https://github.com/lightningnetwork/lnd/blob/master/itest/lnd_multi-hop_test.go) it took me quite some time to maintain while working on the sweeper series and I'm refactoring them now.

I'm also struggling to apply these rules of thumb from time to time because we need to improve the testability of the code. My general take is, as long as the tests are easy to maintain I'm happy. This means if table-driven tests are not used, I'm good, as the tests can always be refactored to use tables (low priority IMO); if table-driven tests are **properly** used, I'm good. I'm against improper usage of table-driven tests which lead to high maintenance costs.

So in **most** of cases, I'd avoid using table-driven tests in itest. I'd also recommend this [Advanced Testing with Go](https://www.youtube.com/watch?v=8hQG7QlcLBk&t=1234s&ab_channel=GopherAcademy) talk, especially the "REPEAT YOURSELF" part.

Finally I would be very hesitant to make changes to any PR, unless critical, if it already got two approvals and no Request-for-change from others🤓